raiden_contracts/tests/test_channel_unlock.py

```diff
@@ -918,6 +924,83 @@ def test_channel_unlock_with_a_large_expiration(
     assert balance_contract == pre_balance_contract - values_B.locked
 
 
+def test_reverse_participants_unlock(
+        web3,
+        token_network,
+        get_accounts,
+        create_settled_channel,
+        reveal_secrets,
+):
+    (A, B, C) = get_accounts(3)
+    settle_timeout = 12
+
+    # Mock pending transfers data
+    pending_transfers_tree_A = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    pending_transfers_tree_B = get_pending_transfers_tree(web3, [3, 4, 6], [4], settle_timeout)
+    assert pending_transfers_tree_A.merkle_root != pending_transfers_tree_B.merkle_root
+
+    # Reveal secrets
+    reveal_secrets(A, pending_transfers_tree_A.unlockable)
+    reveal_secrets(B, pending_transfers_tree_B.unlockable)
+
+    # Settle the channel
+    create_settled_channel(
+        A,
+        pending_transfers_tree_A.locked_amount,
+        pending_transfers_tree_A.merkle_root,
+        B,
+        pending_transfers_tree_B.locked_amount,
+        pending_transfers_tree_B.merkle_root,
+        settle_timeout,
+    )
+
+    # A trying to unlock its own locksroot & locked amount MUST fail
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            A,
+            B,
+            pending_transfers_tree_A.packed_transfers,
+        ).transact({'from': A})
+
+    # Delegate trying to unlock A's own locksroot & locked amount on behalf of A MUST fail
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            A,
+            B,
+            pending_transfers_tree_A.packed_transfers,
+        ).transact({'from': C})
+
+    # B trying to unlock its own locksroot & locked amount MUST fail
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            B,
+            A,
+            pending_transfers_tree_B.packed_transfers,
+        ).transact({'from': B})
+
+    # Delegate trying to unlock B's own locksroot & locked amount on behalf of B MUST fail
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            B,
+            A,
+            pending_transfers_tree_B.packed_transfers,
+        ).transact({'from': B})
+
+    # Someone trying to unlock B's locksroot & locked amount on behalf of A MUST succeed
+    token_network.functions.unlock(
+        A,
+        B,
+        pending_transfers_tree_B.packed_transfers,
+    ).transact({'from': C})
+
+    # Someone trying to unlock A's locksroot & locked amount on behalf of B MUST succeed
+    token_network.functions.unlock(
+        B,
+        A,
+        pending_transfers_tree_A.packed_transfers,
+    ).transact({'from': C})
+
+
```

Comment:
Maybe add
```
token_network.functions.unlock(
        A,
        A,
        pending_transfers_tree_A.packed_transfers,
    ).transact({'from': A})
```

Reply:
I would not add this test here, because `(A, A)` is a different channel than the one we are testing here - `(A, B)` or `(B, A)`, so it would effectively fail at this line of code:
https://github.com/raiden-network/raiden-contracts/blob/903934941f25dcae64bf2d8dc50c3e0882db6a3e/raiden_contracts/contracts/TokenNetwork.sol#L663 instead of testing the `reverse locksroot` issue.

We do have a test for not being able to open a channel `(A, A)` here 
https://github.com/raiden-network/raiden-contracts/blob/903934941f25dcae64bf2d8dc50c3e0882db6a3e/raiden_contracts/tests/test_channel_open.py#L38-L40

But it makes sense to have an additional test for all other functions, to make sure that the transactions will fail + a test specifically for `getChannelIdentifier`.

Reply:
> We do have a test for not being able to open a channel (A, A) here 
It's a slightly different issue than 'reverse locksroot' but related, as the vector is to break the 'participant/partner' authenticity.

Reply:
Added: https://github.com/raiden-network/raiden-contracts/pull/160/commits/7ccfe763c8c8329b367233a3708b472fc530bcda